[JENKINS-35184] Servlet API dependent bits in a inner-class

ydubreuil opened the pull request:
In order to avoid loading `ServletContextListener` class from slaves, this patch moves code depending on the Servlet API into a static inner-class.

It fixes [JENKINS-35184](https://issues.jenkins-ci.org/browse/JENKINS-35184). Impact of the problem is not clear for the general case, it seems to only happen during development (when Jenkins is started with `mvn hpi:run`) in Pipeline jobs.

@reviewbybees esp @oleg-nenashev

ydubreuil committed:
[JENKINS-35184] Servlet API dependent bits in a inner-class

This is in order to avoid loading ServletContextListener class
from slaves classloader.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

stephenc commented:
Hmmmm this says to me that the design of the `SystemProperties` class is invalid and needs reconsideration.

In a correct implementation `SystemProperties` would be an abstract class implementing `Serializable`. The Jenkins JVM would have an implementation that uses the `ServletContextListener` and hooks itself up as the singleton for that `Jenkins` instance. It would also have a `writeReplace()` method that replaces itself with a wrapped `Channel.export()` proxy.

We would need to clarify what the order of property resolution should be on agent JVMs.

In the absence of such a change then it would seem that this class should actually be banned from resolution on the agent JVMs (much like the `Jenkins` class should be banned)

I guess I am :bug: on this change as this is masking actual :bug: s in plugins (i.e. they are serializing classes that are currently designed to work only on the master JVM)

ndeloof commented:
https://i.imgflip.com/1avbid.jpg

ndeloof commented:
Agree there's a lower level design issue here, and/or something weird hasn't been detected yet when running in dev mode with `hpi:run`, anyway this bug just prevent plugin development with 2.4 and later :'(

![1avbid](https://cloud.githubusercontent.com/assets/132757/18631570/062a0a8c-7e73-11e6-97f6-f970ef1f4ad7.jpg)

oleg-nenashev commented:
I'm fine with the approach as a temporary solution.
By the way, I'm not sure it's going to solve the issue completely since <code>SystemProperties#tryGetValueFromContext</code> also depends on the <code>ServletContext</code> class, which may cause similar issues

oleg-nenashev commented:
The design could be reworked, but actually it's valid for the simple cases SystemProperties was designed for. It was just a temporary stub for configuring Jenkins master settings in WAR packages. <code>@Restricted</code> is there for a reason.

IMHO "design is invalid" sounds rather like "not designed as @stephenc wants" (not the first time). @stephenc had a chance to review the pull request, and he should also know timing considerations for this implementation.

jglick reviewed (changes requested):
As @stephenc said, this is not the right fix at all. I think `RingBufferLogHandler` is what needs to be fixed. The no-arg constructor should be deprecated, and `LogRecorder`, `WebAppMain`, and `SlaveComputer` adjusted accordingly.

jglick commented:
For extra credit, implement JENKINS-26677 and demonstrate that it reproduces this problem in a functional test.

ydubreuil commented:
I'm not sure that `SystemProperties` methods are called at all on the slave side. The problem is comes from https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/slaves/SlaveComputer.java#L815 which triggers the class loading to get the class name (btw, I don't understand how `LOGGER.addHandler` can work two lines below, remoting is black magic to me).

I agree with @jglick that fixing `RingBufferLogHandler` is the right thing to do. I'll update the PR.

stephenc commented:
@oleg-nenashev let me try rephrasing then:

*IF* the intended use of the `SystemProperties` class was _only_ from the master JVM *THEN* the only defect in the `SystemProperties` class is that the javadoc comment did not include such a disclaimer *OTHERWISE* it is clear that the class is completely unsuitable for use from an agent JVM and it is exceedingly poorly designed for use from an agent JVM.

Getting code through code review does not absolve that code from additional critiques afterwards, and one should not take a critique of the code personally.

The PR as it currently stands would solidify the use case of calling this API from an agent. *If* we are considering a PR to solidify that use case then I would expect:

1. Clarification of how to resolve conflicts between system properties defined on the agent JVM and system properties defined on the master JVM.
2. Clarification of how to resolve conflicts between the master's servlet context and the agent's JVM
3. An actual remoting safe implementation. This would require defining an API interface and using `Channel.export` when sending the singleton over to an agent JVM.

All of this makes the task significantly harder. The better quick fix is to stop the class being loaded on the remote agent in the first place. Elsewhere we can start progress on blacklisting master only APIs from remote agents (which will need to be turned on in `hpi:run` first and only after some time will we be able to turn it on for production systems)

stephenc commented:
I have created https://github.com/jenkinsci/jenkins/pull/2556 with a view towards starting to mark master-jvm-only classes within Jenkins core and plugins

oleg-nenashev commented:
> IF the intended use of the SystemProperties class was only from the master JVM THEN the only defect in the SystemProperties class is that the javadoc comment did not include such a disclaimer

Exactly. Such comment was definitely required. During the code I've just performed manual check of potential usages on remote side, but no wonder I've missed something

>  Elsewhere we can start progress on blacklisting master only APIs from remote agents (which will need to be turned on in hpi:run first and only after some time will we be able to turn it on for production systems)

Must-have in any case. I could try implementing static analyzer in FindBugs (deep-dive is acceptable for "Max" effort) or BCT for this case.

ydubreuil committed:
[JENKINS-35184] Don't use SystemProperties while initializing remote agents

This rolls back the previous commit and introduces a new way to construct
RingBufferLogHandler which avoids relying on SystemProperties to get the
default size.

ndeloof reviewed (approved):
LGTM

ydubreuil commented:
I updated the PR to avoid calling `SystemProperties` methods in `RingBufferLogHandler` and rolled back changes in `SystemProperties`

I'm not completely confident with the fact that `SLAVE_LOG_HANDLER` is now `null` during the early initialization of the remote agent. My understanding is that it is not used until it is registered by `SlaveInitializer`, so should be safe, but is it really the case?

I was not able to write a test which fails to load `SystemProperties` on the remote agent. It worked all the time. On remote agent,  `servletContext` is obviously null and getting properties defined in the Web application doesn't work. I also have a few tests fixes that I'll push in another PR.

stephenc reviewed (changes requested):
I think this introduces a regression whereby if multiple classloaders are instantiated on the agent we can get multiple copies of the log handler installed

ydubreuil committed:
[JENKINS-35184] Mark SystemProperties as OnMaster only class

Adding `OnMaster` annotation does not prevent the class from being
loaded on remote agent but it gives a hint that this class should not
be used on a remote agent.

ydubreuil commented:
Yeah, you're right  @stephenc , `SLAVE_LOG_HANDLER` will always be null...

ydubreuil committed:
[JENKINS-35184] Set SLAVE_LOG_HANDLER at the very beginning

In the previous code cleaning existing log handlers, SLAVE_LOG_HANDLER
is always null, as static fields are scoped by classloader.

ydubreuil commented:
I think my last commit fixes the regression @stephenc

stephenc reviewed (approved):
:bee:

oleg-nenashev commented:
👎 for merging without giving a chance to the community to review changes

ydubreuil commented:
@oleg-nenashev feel free to rollback if you consider it's better

daniel-beck commented:
@oleg-nenashev So, what are we going to do here?

jglick commented:
retro:bee: from me. If someone has a specific complaint about the fix, they can ask to revert it, or amend it.

oleg-nenashev commented:
@daniel-beck @jglick @ydubreuil 
I see no need in reverting it since the fix is pretty consistent itself. It looks good to me in the current state (would be nice to have some tests BTW).

My only concern was the merge approach
